pkg/util/azureclient/roundtripper.go

```diff
@@ -0,0 +1,81 @@
+package azureclient
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	utillog "github.com/Azure/ARO-RP/pkg/util/log"
+)
+
+// genevaOutboundRequestsTable is the table name we configure in Geneva
+// to send logs of outgoing requests from ARO-RP to ARM.
+// https://docs.google.com/document/d/1RbnKKPNjw7kJZeR-2me4euM5CClGqF62FkaHVlWGxtM/edit?usp=sharing
+const genevaOutboundRequestsTable = "outboundRequests"
+
+func NewCustomRoundTripper(next http.RoundTripper) http.RoundTripper {
+	return &customRoundTripper{
+		next: next,
+	}
+}
+
+type customRoundTripper struct {
+	next http.RoundTripper
+}
+
+func (crt *customRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	correlationData := api.GetCorrelationDataFromCtx(req.Context())
+	if correlationData == nil {
+		correlationData = api.CreateCorrelationDataFromReq(req)
+	}
+
+	requestTime := time.Now()
+	l := updateCorrelationDataAndEnrichLogWithRequest(correlationData, utillog.GetLogger(), requestTime, req)
+
+	l.Info("HttpRequestStart")
+
+	res, err := crt.next.RoundTrip(req)
+
+	l = updateCorrelationDataAndEnrichLogWithResponse(correlationData, l, res, requestTime)
+	l.Info("HttpRequestEnd")
+
+	return res, err
+}
+
+// updateCorrelationDataAndEnrichLogWithRequest receives a non nil correlationData and updates the request time.
+// It also returns a new logrus entry updated with the new field for the LOGKIND entry (custom DGrep table for outgoing requests).
+func updateCorrelationDataAndEnrichLogWithRequest(correlationData *api.CorrelationData, l *logrus.Entry, requestTime time.Time, req *http.Request) *logrus.Entry {
+	correlationData.RequestTime = requestTime
+
+	l = utillog.EnrichWithCorrelationData(l, correlationData)
+	l = l.WithFields(logrus.Fields{
+		"request_URL": req.URL,
+		"LOGKIND":     genevaOutboundRequestsTable,
+	})
+
+	return l
+}
+
+func updateCorrelationDataAndEnrichLogWithResponse(correlationData *api.CorrelationData, l *logrus.Entry, res *http.Response, requestTime time.Time) *logrus.Entry {
+	if res == nil {
+		return l.WithFields(logrus.Fields{
+			"response_status_code":   "0",
+			"contentLength":          "-1",
+			"durationInMilliseconds": time.Since(requestTime).Milliseconds(),
+		})
+	}
+
+	correlationData.CorrelationID = res.Header.Get("X-Ms-Correlation-Request-Id")
```

Comment:
nit: this might be worth adding, or in the `CorrelationData` struct?

https://docs.google.com/document/d/1RbnKKPNjw7kJZeR-2me4euM5CClGqF62FkaHVlWGxtM/edit
>  Technically speaking, the client performs a request to ARM with x-ms-client-request-id, ARM receives the request and in the response, it adds the x-ms-correlation-request-id (the flow is client->ARM->ARO-RP->ARM) but it was simplified in Figure 1.

Reply:
This footnote in the Spike exists because the picture with all the details was too big. This is just a comment to clarify that there should be one more "box" in the picture. As the Spike exists, I didn't want to put many comments in the code to not pollute it. 🙂

Reply:
I included the link to the Spike in pkg/util/azureclient/roundtripper.go to not loose the spike in the ocean of files we have and have it as a reference 🙂

Reply:
This is just a nit so you don't have to reflect it, but I don't think leaving a small comment pollutes the code.
You don't have to explain all your design, but adding comments to some "magic" variables is useful as you did here.

https://github.com/Azure/ARO-RP/blob/4b06345c878558a06d4207db4ea3fd1c651f964f/pkg/util/azureclient/roundtripper.go#L16-L19